component/attr_cache/cacheMap.go

```diff
@@ -55,39 +58,102 @@ func newAttrCacheItem(attr *internal.ObjAttr, exists bool, cachedAt time.Time) *
 		attrFlag: 0,
 		cachedAt: cachedAt,
 	}
-
 	item.attrFlag.Set(AttrFlagValid)
 	if exists {
 		item.attrFlag.Set(AttrFlagExists)
 	}
-
 	return item
 }
 
+func (value *attrCacheItem) insert(attr *internal.ObjAttr, exists bool, cachedAt time.Time) *attrCacheItem {
+	if attr == nil {
+		return nil
+	}
+	path := internal.TruncateDirName(attr.Path)
+	//start recursion
+	cachedItem := value.insertHelper(attr, exists, cachedAt, path, "")
+	return cachedItem
+}
+
+// TODO: write unit tests for this
+func (value *attrCacheItem) insertHelper(attr *internal.ObjAttr, exists bool, cachedAt time.Time, path string, itemPath string) *attrCacheItem {
+	var cachedItem *attrCacheItem
+	paths := strings.SplitN(path, "/", 2) // paths[0] is home paths[1] is user/folder/file
+	if value.children == nil {
+		value.children = make(map[string]*attrCacheItem)
+	}
+	if len(paths) < 2 {
+		// this is a leaf
+		cachedItem = newAttrCacheItem(attr, exists, cachedAt)
+		value.children[paths[0]] = cachedItem
+	} else {
+		itemPath += paths[0] + "/"
+		//see if the directory exists. if not, create it.
+		_, ok := value.children[paths[0]]
+		if !ok {
+			value.children[paths[0]] = newAttrCacheItem(internal.CreateObjAttrDir(itemPath), exists, cachedAt)
+		}
+		cachedItem = value.children[paths[0]].insertHelper(attr, exists, cachedAt, paths[1], itemPath)
+	}
+	return cachedItem
+}
+
+// input: full path to item or file as string
+// output: the attrCacheItem value for the key found in path
+// description: a lookup of any attrCacheItem based on any given full path.
+func (value *attrCacheItem) get(path string) (*attrCacheItem, error) {
+	path = internal.TruncateDirName(path)
+	paths := strings.Split(path, "/")
+	currentItem := value
+	for _, pathElement := range paths {
+		//check if we are at the last element in the paths list
+		if path == "" {
+			break
+		}
+		var ok bool
+		currentItem, ok = currentItem.children[pathElement]
+		//check to see if directory (pathElement) exists
+		if !ok {
+			return nil, fmt.Errorf("The path element : %s does not exist", pathElement)
```

Comment:
```suggestion
			return nil, fmt.Errorf("Cache entry for path %s not found", path)
```
The caller will want an error message with the full path.